src/sisl/cli/sdata.py

```diff
@@ -0,0 +1,370 @@
+    do is to add the method to the data class and make sure it is type annotated. The
+    method should then automatically show up in the CLI.
+
+  - By adding a new type of data. In this case, you need to create the data class and
+    make sure that it is added to the CLI in this file. You need to add a new line to
+    the `main` function that calls `add_data_group` with the new data class.
+
+Alternatively, it is also possible that you might want to change the way the CLI works.
+This is of course more complex, but here is a sketch of how the CLI works to help
+pinpoint where you might want to make changes:
+
+    - The CLI is a click CLI.
+    - The CLI is created on `main`.
+    - The main app is created and then we add a group for each data class. Each group
+      contains:
+        - Commands to generate data:
+          We automatically gather all the possible sources from the `new` dispatcher
+          of the data class. We then decorate them automatically with
+          `decorate_data_gen_function` to be used in click based on type annotations.
+        - Commands to post process data:
+          We automatically gather all the methods of the data class that are not private
+          and decorate them with `decorate_data_method` to be used in click based on type
+          annotations.
+
+    - The way we can generate and post process data in the same CLI is by allowing commands
+      to be chained. However, since commands execute independently, we need a global storage
+      so that the post processing commands can access the generated data. We store the
+      generated data in the `GENERATED` global variable.
+
+    - To decorate functions automatically based on type, we rely on the `decorate_click`
+      function from `nodify`. However, there are some types that are specific to `sisl`. We
+      need to define how those types will be generated from the values (strings) passed in
+      the `CLI`. We do that in the `custom_type_kwargs` dictionary.
+
+"""
+import functools
+import inspect
+import sys
+from typing import Callable, Optional
+
+import rich_click as click
+from nodify.cli._click import decorate_click
+
+import sisl
+from sisl.data import Data, DOSData
+
+# Some configuration to tweak CLICK display.
+click.rich_click.TEXT_MARKUP = "markdown"
+click.rich_click.SHOW_ARGUMENTS = True
+
+# Define all the possible annotations that are custom to sisl and
+# should be supported.
+custom_type_kwargs = {
+    # All sile classes. If a CLI has an argument that is annotated as a sile
+    # then the value provided to the CLI will be parsed simply as sile_cls(value),
+    # i.e. a sile will be created from the user provided path.
+    **{sile_cls: {"parser": sile_cls} for sile_cls in sisl.get_siles()},
+}
+
+# List that keeps track of the data that the CLI has generated.
+# When chaining commands, the first command will generate data and
+# the subsequent commands will be methods to apply on that data.
+GENERATED = []
```

Comment:
I think this is sub-optimal, imagine one parses a grid, and a geometry (from two sources) and then writes them out, then one has to figure out the order.

I think click contexts are better here, with a dict or something, then sub-commands can also change data to other data-types.
I guess we just need some way of distinguishing data, and also enable the possibility of adding more of the same data-type. E.g. comparing two DOS is not that unreasonable, and here it would be, perhaps better, to have a data-container that contains origin of data. So one can query where it came from.